pkg/asset/cluster/openstack/bootstrapshim.go

```diff
@@ -0,0 +1,228 @@
+func buildBootstrapShim(ctx context.Context, cloud string, infraID string, proxy *types.Proxy, bootstrapIgn []byte) ([]byte, error) {
+	opts := openstackdefaults.DefaultClientOpts(cloud)
+	conn, err := openstackdefaults.NewServiceClient("image", opts)
+	if err != nil {
+		return nil, err
+	}
+
+	var userCA []byte
+	{
+		cloudConfig, err := clientconfig.GetCloudFromYAML(opts)
+		if err != nil {
+			return nil, err
+		}
+		// Get the ca-cert-bundle key if there is a value for cacert in clouds.yaml
+		if caPath := cloudConfig.CACertFile; caPath != "" {
+			userCA, err = os.ReadFile(caPath)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read clouds.yaml ca-cert from disk: %w", err)
+			}
+		}
+	}
+
+	var glancePublicURL string
+	{
+		var serviceCatalog *tokens.ServiceCatalog
+		{
+			authResult := conn.GetAuthResult()
+			auth, ok := authResult.(tokens.CreateResult)
+			if !ok {
+				return nil, fmt.Errorf("unable to extract service catalog")
+			}
+
+			var err error
+			serviceCatalog, err = auth.ExtractServiceCatalog()
+			if err != nil {
+				return nil, err
+			}
+		}
+		clientConfigCloud, err := clientconfig.GetCloudFromYAML(openstackdefaults.DefaultClientOpts(cloud))
+		if err != nil {
+			return nil, err
+		}
+		glancePublicURL, err = gophercloud_openstack.V3EndpointURL(serviceCatalog, gophercloud.EndpointOpts{
+			Type:         "image",
+			Availability: gophercloud.AvailabilityPublic,
+			Region:       clientConfigCloud.RegionName,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("cannot retrieve Glance URL from the service catalog: %w", err)
+		}
+	}
+
+	// upload the bootstrap Ignition config in Glance and save its location
+	var bootstrapConfigURL string
+	{
+		img, err := images.Create(conn, images.CreateOpts{
+			Name:            infraID + "-ignition",
+			ContainerFormat: "bare",
+			DiskFormat:      "raw",
+			Tags:            []string{"openshiftClusterID=" + infraID},
+		}).Extract()
+		if err != nil {
+			return nil, fmt.Errorf("unable to create a Glance image for the bootstrap server's Ignition file: %w", err)
+		}
+
+		if res := imagedata.Upload(conn, img.ID, bytes.NewReader(bootstrapIgn)); res.Err != nil {
+			return nil, fmt.Errorf("unable to upload a Glance image for the bootstrap server's Ignition file: %w", res.Err)
+		}
+
+		bootstrapConfigURL = glancePublicURL + img.File
+	}
+
+	// To allow Ignition to download its config on the bootstrap machine from a location secured by a
+	// self-signed certificate, we have to provide it a valid custom ca bundle.
+	// To do so we generate a small ignition config that contains just Security section with the bundle
+	// and later append it to the main ignition config.
+	// We can't do it directly in Terraform, because Ignition provider suppors only 2.1 version, but
+	// Security section was added in 2.2 only.
+	// generateIgnitionShim is used to generate an ignition file that contains a user ca bundle
```

Comment:
leftover?

Reply:
not really; this code is used in the Terraform flow as well.
But as you point out, it will soon be. I'm deleting these two lines, thanks for the comment.

Reply:
oh wait you meant the `generateIgnitionShim` part. Yes it is! you are correct. Removing that as well.